File: resources/android/StatusLabelControl.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.compositeOver
import androidx.compose.ui.graphics.luminance
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.nativephp.plugins.native_ui.NativeUITokens

enum class StatusLabelTone(val wireName: String) {
    Neutral("neutral"),
    Info("info"),
    Success("success"),
    Warning("warning"),
    Danger("danger");

    companion object {
        fun fromWire(value: String): StatusLabelTone = entries.firstOrNull {
            it.wireName == value
        } ?: Neutral
    }
}

data class StatusLabelTokenColors(
    val background: Color,
    val foreground: Color,
)

internal fun resolveStatusLabelTokenColors(
    tokens: NativeUITokens,
    tone: StatusLabelTone,
): StatusLabelTokenColors {
    val (candidateBackground, candidateForeground) = when (tone) {
        StatusLabelTone.Neutral -> tokens.surfaceVariant to tokens.onSurfaceVariant
        StatusLabelTone.Info -> tokens.primary to tokens.onPrimary
        StatusLabelTone.Success -> tokens.success to tokens.onSuccess
        StatusLabelTone.Warning -> tokens.accent to tokens.onAccent
        StatusLabelTone.Danger -> tokens.destructive to tokens.onDestructive
    }
    val background = candidateBackground.compositeOver(tokens.surface).copy(alpha = 1f)
    val preferredForeground = candidateForeground.compositeOver(background).copy(alpha = 1f)
    val foreground = preferredForeground.takeIf {
        statusLabelContrastRatio(it, background) >= 4.5f
    } ?: listOf(Color.Black, Color.White).maxBy {
        statusLabelContrastRatio(it, background)
    }

    return StatusLabelTokenColors(background, foreground)
}

internal fun firstlightStatusLabelDescription(
    label: String,
    accessibilityLabel: String,
    accessibilityHint: String,
): String = listOf(accessibilityLabel.ifEmpty { label }, accessibilityHint)
    .filter(String::isNotEmpty)
    .joinToString(". ")

@Composable
fun FirstlightStatusLabelControl(
    label: String,
    colors: StatusLabelTokenColors,
    accessibilityDescription: String,
    modifier: Modifier = Modifier,
) {
    Surface(
        modifier = modifier.semantics(mergeDescendants = true) {
            contentDescription = accessibilityDescription
        },
        shape = RoundedCornerShape(percent = 50),
        color = colors.background,
        contentColor = colors.foreground,
    ) {
        Text(
            text = label,
            modifier = Modifier.padding(horizontal = 10.dp, vertical = 5.dp),
            color = colors.foreground,
            style = MaterialTheme.typography.labelMedium,
            fontWeight = FontWeight.SemiBold,
        )
    }
}

internal fun statusLabelContrastRatio(foreground: Color, background: Color): Float {
    val lighter = maxOf(foreground.luminance(), background.luminance())
    val darker = minOf(foreground.luminance(), background.luminance())
    return (lighter + 0.05f) / (darker + 0.05f)
}

File: resources/ios/StatusLabelControl.swift
import SwiftUI
import UIKit

enum StatusLabelTone: String, CaseIterable {
    case neutral
    case info
    case success
    case warning
    case danger
}

struct FirstlightStatusLabelTokens {
    let background: UIColor
    let foreground: UIColor

    static func from(
        theme: NativeUITokens,
        tone: StatusLabelTone,
        traits: UITraitCollection
    ) -> Self {
        let pair: (background: Color, foreground: Color) = switch tone {
        case .neutral: (theme.surfaceVariant, theme.onSurfaceVariant)
        case .info: (theme.primary, theme.onPrimary)
        case .success: (theme.success, theme.onSuccess)
        case .warning: (theme.accent, theme.onAccent)
        case .danger: (theme.destructive, theme.onDestructive)
        }

        let surface = statusLabelOpaqueComposite(
            foreground: UIColor(theme.surface),
            over: .systemBackground,
            traits: traits
        )
        let background = statusLabelOpaqueComposite(
            foreground: UIColor(pair.background),
            over: surface,
            traits: traits
        )
        let preferred = statusLabelOpaqueComposite(
            foreground: UIColor(pair.foreground),
            over: background,
            traits: traits
        )
        let foreground = statusLabelContrastRatio(preferred, background) >= 4.5
            ? preferred
            : [UIColor.black, .white].max {
                statusLabelContrastRatio($0, background)
                    < statusLabelContrastRatio($1, background)
            }!

        return Self(background: background, foreground: foreground)
    }
}

struct FirstlightStatusLabelControl: View {
    let label: String
    let tokens: FirstlightStatusLabelTokens
    let accessibilityLabel: String
    let accessibilityHint: String

    let isInteractive = false

    var body: some View {
        Text(label)
            .font(.footnote.weight(.semibold))
            .foregroundStyle(Color(uiColor: tokens.foreground))
            .lineLimit(nil)
            .fixedSize(horizontal: false, vertical: true)
            .padding(.horizontal, 10)
            .padding(.vertical, 5)
            .background(
                Capsule().fill(Color(uiColor: tokens.background))
            )
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(Text(accessibilityLabel))
            .accessibilityHint(Text(accessibilityHint))
    }
}

func statusLabelContrastRatio(_ foreground: UIColor, _ background: UIColor) -> CGFloat {
    let foregroundLuminance = statusLabelRelativeLuminance(foreground)
    let backgroundLuminance = statusLabelRelativeLuminance(background)

    return (max(foregroundLuminance, backgroundLuminance) + 0.05)
        / (min(foregroundLuminance, backgroundLuminance) + 0.05)
}

private func statusLabelOpaqueComposite(
    foreground: UIColor,
    over background: UIColor,
    traits: UITraitCollection
) -> UIColor {
    let foreground = statusLabelRGBA(foreground, traits: traits)
    let background = statusLabelRGBA(background, traits: traits)
    let alpha = foreground.alpha + background.alpha * (1 - foreground.alpha)

    guard alpha > 0 else { return .black }

    return UIColor(
        red: (foreground.red * foreground.alpha
            + background.red * background.alpha * (1 - foreground.alpha)) / alpha,
        green: (foreground.green * foreground.alpha
            + background.green * background.alpha * (1 - foreground.alpha)) / alpha,
        blue: (foreground.blue * foreground.alpha
            + background.blue * background.alpha * (1 - foreground.alpha)) / alpha,
        alpha: alpha
    )
}

private func statusLabelRGBA(
    _ color: UIColor,
    traits: UITraitCollection = .current
) -> (red: CGFloat, green: CGFloat, blue: CGFloat, alpha: CGFloat) {
    let resolved = color.resolvedColor(with: traits)
    var red: CGFloat = 0
    var green: CGFloat = 0
    var blue: CGFloat = 0
    var alpha: CGFloat = 0

    guard resolved.getRed(&red, green: &green, blue: &blue, alpha: &alpha) else {
        return (0, 0, 0, 1)
    }

    return (red, green, blue, alpha)
}

private func statusLabelRelativeLuminance(_ color: UIColor) -> CGFloat {
    let components = statusLabelRGBA(color)

    func linearized(_ channel: CGFloat) -> CGFloat {
        channel <= 0.04045
            ? channel / 12.92
            : pow((channel + 0.055) / 1.055, 2.4)
    }

    return 0.2126 * linearized(components.red)
        + 0.7152 * linearized(components.green)
        + 0.0722 * linearized(components.blue)
}
